### ISSUE TITLE
seaweedfs 2.84 - weed fix would be better if it just took .dat file paths as parameter

### ISSUE THREAD

**Issue by Radtoo** [chrislusf/seaweedfs#2563]
It think for larger collections it would be easier to have weed fix just operate on paths to dat files. Then you could just use a more expedient and intuitive `weed fix *.dat` or `weed fix /path/**/*.dat` or `weed fix /path/*collectionname*.dat`. Or a variant with commonly used tools like find, fd or such.

I believe right now if you have a lot of files and maybe directories that need "weed fix" (e.g. to migrate to a large_disk variant or because something went wrong), you have have to script the command to be fed a number sequence generator (e.g. "seq") and maybe handle errors arising from volumes absent from this machine/directory (needed for example with xargs), or alternatively look at the file names and parse both volume numbers and collection names to be passed to weed fix.  It just seems a lot harder.

**Comment by chrislusf** [comment]
the logic is all in this one file: https://github.com/chrislusf/seaweedfs/blob/master/weed/command/fix.go

If you want to learn some go skills, this would be a good opportunity. :)

**Comment by Radtoo** [comment]
> the logic is all in this one file: https://github.com/chrislusf/seaweedfs/blob/master/weed/command/fix.go
> 
> If you want to learn some go skills, this would be a good opportunity. :)

Thanks for the hint! I did manage to get this far https://github.com/Radtoo/seaweedfs/blob/fix_paths/weed/command/fix.go. It compiles and works fine when I try it on my Linux CLI (colle*.dat, 6*.dat, /path/to/collection_*.dat and so on - much faster to work with). 

But the code might not actually be good style or excellent use of the standard lib/syntax with my _whole_ few hours of Go experience. Also I'm not sure about the (non-?)verbosity of the output logged, Would you be willing to give it a look?

**Comment by chrislusf** [comment]
Nice progress!

* Use named parameters, similar to this one https://github.com/Radtoo/seaweedfs/blob/fix_paths/weed/command/scaffold.go#L32
* Refactor a bit to move logic of one volume to its own function.

**Comment by Radtoo** [comment]
> Use named parameters, similar to this one https://github.com/Radtoo/seaweedfs/blob/fix_paths/weed/command/scaffold.go#L32

I couldn't find out how it would assign a variable name to the default positional arguments via .Flag or if/why that even makes sense.

Or do you actually want something surprising like requiring an option flag like weed fix -datFilesToFix=colle*.dat,6*.dat,/path/to/collection_*.dat or -datFilesToFix="colle*.dat 6*.dat /path/to/collection_*.dat"? Seems strange for _required_ files/paths that aren't actually options, and is also not done on for example weed upload which equally just specifies files as positional arguments.

> Refactor a bit to move logic of one volume to its own function.
Hm, I don't believe you just want the inside of that one top level loop that handles 1 to n arguments as a function that instead handles just strictly 1, right? It is anyhow only running once if there is only one argument.

The other code handling the actual volumes was essentially already there and I don't understand it in detail to refactor it, I just hand it the same data parsed from paths rather than three options passed to weed fix running once for each volume.

**Comment by chrislusf** [comment]
I added my fix 110d5a52337cb6c9d8684a5fd453fbb5c4214b41

Does it work for you?

**Comment by Radtoo** [comment]
> Does it work for you?

It wouldn't quite have worked with 110d5a5, because shell usage like "weed fix /mnt/iscsi*/weed/**/*_42.dat" or "fd --changed-within 2w -x weed fix" is not supported. 

I did change your new version of fix to be like this 389002f19509a380bbc97ebd01e39395689b6c79 which feels natural to use via shell. But of course it again uses args.

One commit before that (fba1efb77a9cbb620eab427c26a8983042f493ed) I tried to use a -dir (-path) type option like you seem to prefer  but I actually can't figure out how Go's "Flag" can be told to parse results of shell commands like the above to []string when passed to such a -path option.
